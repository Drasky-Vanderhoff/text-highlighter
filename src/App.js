import React, { Component } from 'react'
import { connect } from 'react-redux'
import { clearSelections, saveSelectedText } from 'actions'
import InteractiveColorRow from 'components/InteractiveColorRow'
import ColorTextPresenter from 'components/ColorTextPresenter'
import TextHighlightableArea from 'components/TextHighlightableArea'
import config from './config.json'
import './App.scss'

class App extends Component {  
  constructor(props) {
    super(props)
    // Finds a color that is highlighted in the configurations and set that one as the active one
    const highlightedColor = config.availableColorPairs.find(
      ({isHighlighted}) => isHighlighted
    ).background || ""

    this.state = {
      currentText: config.text,
      selectionColor: highlightedColor,
      filterColor: [highlightedColor]
    }
  }

  /* Callbacks */
  saveSelectedText = (selectionStart, selectionEnd) => this.props.saveSelectedText(
    this.state.selectionColor,
    selectionStart,
    selectionEnd,
    this.state.currentText
  )
  setSelectionColor = ([selectionColor]) => this.setState({selectionColor})
  setFilterColor = (filterColor) => this.setState({filterColor})
  setCurrentText = (currentText) => {
    this.props.clearSelections()
    this.setState({currentText})
  }

  /* Renders */
  renderNoColorsAvailable = () => <div>
    No colors were defined in the configuration, please define colors in order be able to use the app.
  </div>

  render() {
    const filteredSelections = this.props.higlightedTextItems.filter(
      ({color}) => this.state.filterColor.includes(color)
    )
    return (
      <div className="app">
        <button
          data-testid="clear-button" 
          className="clear-button"
          onClick={this.props.clearSelections}
        >
          Clear Selections
        </button>
        {config.availableColorPairs.length ? 
          <InteractiveColorRow
          key="selection_color_row" 
            colorBoxes={config.availableColorPairs} 
            onSelectedColorRow={this.setSelectionColor}
            descriptionText={config.selectorDescriptionText}
            multipleSelections={false}
          /> : 
          this.renderNoColorsAvailable()
        }
        <TextHighlightableArea
          initialText={this.state.currentText}
          onSelectedText={this.saveSelectedText}
          onChangeText={this.setCurrentText}
          highlightMarkers={this.props.higlightedTextItems}
        />
        <InteractiveColorRow
          key="filter_color_row" 
          colorBoxes={config.availableColorPairs} 
          onSelectedColorRow={this.setFilterColor} 
          descriptionText={config.filterDescriptionText}
          multipleSelections={true}
        />
        <ColorTextPresenter colorTextItemsLists={filteredSelections} />
      </div>
    )
  }
}

const mapStateToProps = state => ({
  higlightedTextItems: state.higlightedTextItems
})

const mapDispatchToProps = dispatch => ({
  saveSelectedText: (
    ...params
  ) => dispatch(saveSelectedText(
    ...params
  )),
  clearSelections: () => dispatch(clearSelections())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
