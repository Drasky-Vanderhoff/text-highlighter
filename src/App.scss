.clear-button{
  margin-left: 5%;
}