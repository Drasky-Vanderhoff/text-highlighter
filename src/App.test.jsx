import React from 'react'
import App from './App'
import renderer from 'react-test-renderer'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import rootReducer from './reducers'

const store = createStore(
  rootReducer
)

describe('App Component', () => {
  test('should render with default values', () => {
    expect(
      renderer.create(
        <Provider store={store}>
          <App />
        </Provider>
      ).toJSON()
    ).toMatchSnapshot()
  })
})
