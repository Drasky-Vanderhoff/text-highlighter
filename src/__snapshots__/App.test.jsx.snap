// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render with default values 1`] = `
<div
  className="app"
>
  <button
    className="clear-button"
    data-testid="clear-button"
    onClick={[Function]}
  >
    Clear Selections
  </button>
  <ul
    className="interactive-color-box-row"
  >
    <li
      className="item"
      data-testid="color-box-row-0"
      onClick={[Function]}
    >
      <div
        className="color-box highlighted"
        data-testid="color-box-red"
        style={
          Object {
            "backgroundColor": "red",
            "borderColor": "green",
          }
        }
      />
    </li>
    <li
      className="item"
      data-testid="color-box-row-1"
      onClick={[Function]}
    >
      <div
        className="color-box unhighlighted"
        data-testid="color-box-green"
        style={
          Object {
            "backgroundColor": "green",
            "borderColor": "blue",
          }
        }
      />
    </li>
    <li
      className="item"
      data-testid="color-box-row-2"
      onClick={[Function]}
    >
      <div
        className="color-box unhighlighted"
        data-testid="color-box-blue"
        style={
          Object {
            "backgroundColor": "blue",
            "borderColor": "red",
          }
        }
      />
    </li>
    <li>
      (on the left there is the text market. If a user presses red, he is able highlight the words in the text below in red)
    </li>
  </ul>
  <div
    className="text-highlightable-area"
  >
    <div
      className="backdrop"
    >
      <div
        className="highlights"
        data-testid="highlights"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry’s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially.
      </div>
    </div>
    <textarea
      data-testid="text-area"
      onChange={[Function]}
      onClick={[Function]}
      onTouchEnd={[Function]}
      value="Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry’s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially."
    />
  </div>
  <ul
    className="interactive-color-box-row"
  >
    <li
      className="item"
      data-testid="color-box-row-0"
      onClick={[Function]}
    >
      <div
        className="color-box highlighted"
        data-testid="color-box-red"
        style={
          Object {
            "backgroundColor": "red",
            "borderColor": "green",
          }
        }
      />
    </li>
    <li
      className="item"
      data-testid="color-box-row-1"
      onClick={[Function]}
    >
      <div
        className="color-box unhighlighted"
        data-testid="color-box-green"
        style={
          Object {
            "backgroundColor": "green",
            "borderColor": "blue",
          }
        }
      />
    </li>
    <li
      className="item"
      data-testid="color-box-row-2"
      onClick={[Function]}
    >
      <div
        className="color-box unhighlighted"
        data-testid="color-box-blue"
        style={
          Object {
            "backgroundColor": "blue",
            "borderColor": "red",
          }
        }
      />
    </li>
    <li>
      (on the left there is the filter. If a user presses on red, all red text should appear in the box below)
    </li>
  </ul>
  <blockquote
    className="color-text-presenter"
  />
</div>
`;
